### DISCUSSION TITLE
Problems on Deploy

### DISCUSSION THREAD

**Discussion by carlosevi94** [miguelgrinberg/python-socketio#960] [Q&A]
Hello!

I'm trying to deploy my Django project on a ubuntu server for prepare the production version. But deploy its beeing so complicate. I have few experience deploying normal Django apps, but no experience with complex apps likes app with websockets or something. 

On my local enviroment I have django with eventlet for capture the disconnect events on real time (its important for this project). For server side, first I tried with nginx + gunicorn but gunicorn throw a lots of problems, so I changed to nginx + django runserver.

I'm going to **describe my problem**: 

When I deploy the code in the server, websocket no works, only polling request. Actually I don't know if this is caused by a wrong server configuration or a client problem. 

When I use this method for connect `var socket = io.connect()` and I start a socket.io client connection, on my firefox shell, I see this error

```
Firefox is unable to establish a connection to the server in ws://my_ip_server/socket.io/?EIO=4&transport=websocket&sid=t_elLq2J0SDP6kweAAAA.  websocket.js:54:26
```

Polling request works and return `200 OK` messages. On my development enviroment I don't see that error. 

But if I use this method `var socket = io.connect('http://my_ip_server', { transports: ["polling"] });` I don't see the error above (logicaly, I'm specifying to not use websocket) and the app works like on my local, but disconnect event is not catched on realtime (logicaly too).

I didn't find so many info about nginx + django + socket.io server deployment info, so I used a basic config for start. My nginx conf for this site is this

```
server {
    listen 80;
    server_name XXXXXX; 

   location / {
	 proxy_pass http://127.0.0.1:8000;
    }
}
```

I tryed too this another conf but the result is the same.  
```bash
server {
    listen 80;
    server_name XXXXXX; 

   location / {
	 proxy_pass http://127.0.0.1:8000;
    }

   location /socket.io/ {
	 proxy_pass http://127.0.0.1:8000/socket.io/;
    }
}
```

I don't know if socket.io need another ports open and that is problem. I don't think it because I specified the port on wsgi.py with this lane

```python
eventlet.wsgi.server(eventlet.listen(('127.0.0.1', 8000)), application)
```

What could be the problem?? 

Thanks for read!

### versions 

SO Server: Ubuntu 22.04
SO Local: Ubuntu 20.04

Server Python 3.8.13
Local Python: 3.8.10

```
Package            Version
------------------ -------
asgiref            3.5.2
Django             4.0.5
dnspython          1.16.0
eventlet           0.33.0
greenlet           1.1.2
gunicorn           20.1.0
python-dateutil    2.8.2
python-engineio    4.3.2
python-socketio    5.6.0
```

And the the 4.5.0 of socket.io for js client.

**Comment by miguelgrinberg** [comment]
@carlosevi94 Are you reading the [python-socketio documentation](https://python-socketio.readthedocs.io/en/latest/server.html#deployment-strategies)? There is a lot of information regarding deployment there that you appear to not know about. I also recommend you look at the [Flask-SocketIO documentation](https://flask-socketio.readthedocs.io/en/latest/deployment.html#using-nginx-as-a-websocket-reverse-proxy), which includes an example nginx configuration file, which will work exactly the same for your Django project.

Finally, my understanding is that `django runserver` is not supposed to be used in production. Be careful there, I recommend you go back to Gunicorn. See the docs referenced above for instructions.

**Reply by carlosevi94** [reply]
@miguelgrinberg Thanks for answer!!!

Using your nginx configuration file on Flask-SocketIO documentation  make it works.  
In fact I only need this lines on my nginx configuration file, in `location` section
```
      proxy_http_version 1.1;
      proxy_set_header Upgrade $http_upgrade;
      proxy_set_header Connection "upgrade";
```
But now I'm adapting my nginx conf file for get the best configuration as posible. Thanks!!

Thank for your recomendation. My intention is use gunicorn, but I have a lot of problems and I preferred to go back and go step by step to a stable version. Now I'm having another gunicorn problems (yesterday write on a question that describe the same problem than me).
